doc/poor_mans_mmp_design.txt

```diff
@@ -0,0 +1,70 @@
+MMP (Multi-Modifier Protection) attempts to prevent a user from
+importing a pool on more than one node at a time.
+
+See
+	https://github.com/zfsonlinux/zfs/issues/745
+and
+	https://github.com/zfsonlinux/zfs/pull/5821
+for more background.
+
+This design is for a very simple subset of MMP that can be coded,
+tested, and reviewed quickly.
+
+It attempts to detect and prevent the case of importing a pool via "zpool
+import -f", while it is already imported on another node.
+
+On-disk changes:
+
+The last uberblock slot in each ring is used only for MMP; uberblocks
+written by the sync thread always go into the first (N-1) slots.  The
+slot used for MMP is used to hold an uberblock which is exactly the
+same as the last written uberblock, except that its ub_timestamp is
+frequently updated, and its embedded checksum is different
+accordingly.
+
+This is used to provide the "heartbeat" without overwriting good
+uberblocks.
+
+During Import:
+
+zpool_do_import() (user space)
+1. Examine the hostid.  If it is 0,
+   skip the steps below and perform the import as normal.
+2. Issue the ZFS_IOC_POOL_TRYIMPORT ioctl
+3. Extract the txg and timestamp from the returned config nvlist and
+   save them.
+4. Sleep 1 second.
```

Comment:
Richard, MMP is designed specifically for HA use cases where there are multiple nodes accessing the storage.  We have seen numerous cases in the past where the HA software has failed for one reason or another (e.g. STONITH power control fails, bad HA scripts breaking SCSI reservations and forced import on both nodes, operator error, etc).

While MMP is not in itself an HA exclusion mechanism that can avoid all potential problems, having an independent mechanism at the filesystem level that will sanity check any requests from the HA scripts is definitely beneficial.

I'm not saying that someone should be forced to use MMP, but it is often a case of "I didn't know I needed it" when people need it the most (e.g. VM images, SAN/HA environments where HA is misconfigured or non-existent, etc).  This is expecially true for ZFS, where corruption of the uberblocks and root of the metadata tree is totally fatal for the whole pool.